In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/Library_Usage.csv",low_memory=False)

In [3]:
# Beispiel Mittelwert
# function mean() returns the mean of the values over the requested axis
df['Total Checkouts'].mean()

np.float64(155.218052709061)

In [4]:
# Beispiel Median
# function median() returns the median of the values over the requested axis
df['Total Checkouts'].median()

7.0

##### Aufgabe 3.2 
Schau Dir den Mittelwert und den Median der Variable Total Checkouts an. 
Warum sind die beiden Werte so unterschiedlich?
Was ziehst Du daraus für Schlüsse für weitere statistische Analysen und Reports?

Zur Berechnung des Medians werden zuerst alle Werte in der Spalte "Total Checkouts" in aufsteigender Reihenfolge geordnet.
Danach wird den Wert bestimmt, der die Series genau in zwei Teilen teilt. 
Hat die Series eine gerade Anzahl von Beobachtungen/Ausprägungen, werden dann die zwei Werte, die die Series genau in zwei Teilen teilt, 
addiert und durch 2 geteilt. Das Ergebnis ist der Median. 
Der Median sagt aus, welcher Wert genau in der Mitte einer Datenreihe liegt. 

Zur Berechnung des Mittelwerts werden die einzelen Ausprägungen mit der Anzahl deren Vorkommens multipliziert und durch die Anzahl 
der Beobachtungen geteilt. Das Ergebnis ist der Mittelwert.
    
Für statistische Analyse und Reports ist der Mittelwert aussagekräftig. 
Im Fall der Gesamtausleihen sagt der Mittelwert aus, dass im Durchschnitt 155 Medien ausgeliehen wurden.

##### Quantile

In [5]:
df['Total Checkouts'].quantile(q=[0.25, 0.5, 0.75])

0.25     0.0
0.50     7.0
0.75    74.0
Name: Total Checkouts, dtype: float64

In [6]:
# Ausreißer in einer entfernen bzw. ersetzen
# zuerst die kleinsten oder die größten Beobachtungen identifizieren
# identifies 0.5% of the data at both ends of the distribution
alpha = 0.005
df['Total Checkouts'].quantile([alpha, 1-alpha])

0.005       0.0
0.995    3255.0
Name: Total Checkouts, dtype: float64


##### Aufgabe 3.3 - Ausreißerentfernung

1. Identifziere jeweils die 1.5% größten Werte in der Spalte Total Checkouts. Definiere diese Werte als Ausreißer

In [7]:
alpha = 0.015

In [8]:
df['Total Checkouts'].quantile([alpha, 1-alpha])

0.015       0.0
0.985    1922.0
Name: Total Checkouts, dtype: float64

In [9]:
ausreisser = 0.0

Outliers are extreme
An outlier is an observation that is extreme relative to the rest of the data
TIP: Why it is important to look for outliers
Examination of data for possible outliers serves many useful purposes, including
1. Identifying strong skew in the distribution
2. Identifying data collection or entry errors
3. Providing insight into interesting properties of the data.

In [13]:
df_ausreisser = df.loc[df['Total Checkouts'] == 0.0]

In [14]:
df_ausreisser

Patron Type Definition  Total Checkouts  Total Renewals  \
1                       Adult                0               0   
2                       Adult                0               0   
4                      Senior                0               0   
5                      Senior                0               0   
6                       Adult                0               0   
...                       ...              ...             ...   
450354    Digital Access Card                0               0   
450355    Digital Access Card                0               0   
450356    Digital Access Card                0               0   
450357    Digital Access Card                0               0   
450358    Digital Access Card                0               0   

                Age Range Home Library Definition Circulation Active Month  \
1          45 to 54 years                    Main                      Jul   
2          55 to 59 years        Western Addition                      Mar   
4          65 to 74 years                  Sunset                      Mar   
5       75 years and over                    Main                      Apr   
6          20 to 24 years                    Main                      Feb   
...                   ...                     ...                      ...   
450354     35 to 44 years               Ingleside                     Null   
450355     45 to 54 years               Ingleside                      Aug   
450356     25 to 34 years               Ingleside                      Aug   
450357     35 to 44 years               Ingleside                      Apr   
450358     65 to 74 years                 Bayview                      Apr   

       Circulation Active Year Notice Preference Definition  \
1                         2023                        Email   
2                         2024                        Email   
4                         2024                        Print   
5                         2021                        Email   
6                         2024                        Email   
...                        ...                          ...   
450354                    Null                        Email   
450355                    2022                        Print   
450356                    2022                        Print   
450357                    2022                        Print   
450358                    2023                        Email   

        Provided Email Address  Year Patron Registered  \
1                         True                    2019   
2                         True                    2022   
4                        False                    2023   
5                         True                    2009   
6                         True                    2023   
...                        ...                     ...   
450354                   False                    2023   
450355                   False                    2022   
450356                   False                    2022   
450357                   False                    2022   
450358                   False                    2023   

       Within San Francisco County  
1                            False  
2                            False  
4                            False  
5                            False  
6                            False  
...                            ...  
450354                        Null  
450355                        Null  
450356                        Null  
450357                        Null  
450358                        Null  

[147115 rows x 11 columns]

2. Erstelle einen Datensatz, für den diese Ausreißer entfernt sind.

In [16]:
df_neu = df.loc[df['Total Checkouts'] != 0.0]

In [17]:
df_neu

Patron Type Definition  Total Checkouts  Total Renewals  \
0                      Senior                5               0   
3                     Welcome                1               1   
7                       Adult                1               0   
8                       Adult                4               4   
10                      Adult               24              53   
...                       ...              ...             ...   
450289    Digital Access Card              270             430   
450293    Digital Access Card                1               0   
450297    Digital Access Card                1               0   
450309    Digital Access Card                1               1   
450349    Digital Access Card               71               4   

                Age Range Home Library Definition Circulation Active Month  \
0       75 years and over                    Main                      Nov   
3          20 to 24 years                Richmond                      Aug   
7          10 to 19 years                    Main                      Nov   
8          25 to 34 years                    Main                      Feb   
10         35 to 44 years                 Portola                      Nov   
...                   ...                     ...                      ...   
450289               Null                    Main                      Feb   
450293     10 to 19 years                 Portola                      Jun   
450297     25 to 34 years             North Beach                      Jan   
450309     65 to 74 years               Chinatown                      Mar   
450349     25 to 34 years                    Park                      Dec   

       Circulation Active Year Notice Preference Definition  \
0                         2022                        Email   
3                         2022                        Email   
7                         2022                        Phone   
8                         2024                        Email   
10                        2023                        Email   
...                        ...                          ...   
450289                    2024                        Email   
450293                    2023                         Null   
450297                    2023                        Print   
450309                    2023                        Email   
450349                    2023                        Email   

        Provided Email Address  Year Patron Registered  \
0                         True                    2015   
3                         True                    2022   
7                         True                    2022   
8                         True                    2022   
10                        True                    2019   
...                        ...                     ...   
450289                    True                    2017   
450293                   False                    2023   
450297                   False                    2021   
450309                    True                    2023   
450349                    True                    2013   

       Within San Francisco County  
0                            False  
3                            False  
7                            False  
8                            False  
10                           False  
...                            ...  
450289                        Null  
450293                        Null  
450297                        Null  
450309                        Null  
450349                        Null  

[303244 rows x 11 columns]

3. Handelt es sich hierbei um eine gute Methode, Ausreißer zu identifizieren und zu behandeln? Welche anderen Strategien kennst Du?

The median and IQR are called robust estimates because extreme observations have little effect on their values. 
The mean and standard deviation are much more affected by changes in extreme observations.
The median and IQR are only sensitive to numbers near Q1, the median, and Q3.
Since values in these regions are relatively stable – there aren’t large jumps between observations – the median and IQR estimates are also quite stable.
The interquartile range (IQR, for short) is a measure of variability in data ( like the standard deviation).
The more variable the data, the larger the standard deviation and IQR.
The two boundaries of the box plot are called the first quartile (the 25th percentile, i.e. 25% of the data fall below this value) 
and the third quartile (the 75th percentile), and these are often labeled Q1 and Q3, respectively.

The IQR is the length of the box in a box plot. It is computed as IQR = Q3 − Q1 where Q1 and Q3 are the 25th and 75th percentiles